Added recipe for Inspector Plugin (#48)
godot-rust/gdext-book: src/recipes/editor-plugin/inspector-plugins.md

<!--
  ~ Copyright (c) godot-rust; Bromeon and contributors.
  ~ This Source Code Form is subject to the terms of the Mozilla Public
  ~ License, v. 2.0. If a copy of the MPL was not distributed with this
  ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->

# Inspector plugins

The inspector dock allows you to create custom widgets to edit properties through plugins.
This can be beneficial when working with custom datatypes and resources, although you can
use the feature to change the inspector widgets for built-in types. You can design custom
controls for specific properties, entire objects, and even separate controls associated

> with particular datatypes.

Reviewer: The allowed line width is much more (around 140 chars). You don't have to fully use it, but you can use a bit longer lines than here :slightly_smiling_face:

Author: 150 actually, but I don't really understand why it's so important

so should I change it?

Reviewer: It's not too important, just helps consistency if it doesn't vary wildly between authors. So would be nice if it's not much work, but otherwise feel free to leave it.

For more info, see
[docs.godotengine.org](https://docs.godotengine.org/en/stable/classes/class_editorinspectorplugin.html#class-editorinspectorplugin).

The [example](https://docs.godotengine.org/en/stable/tutorials/plugins/editor/inspector_plugins.html) in the Godot docs in Rust.

Before:

![Before](./images/before.png)

After:

> ![After](./images/after.png)

Reviewer: The first line is not a full sentence.

Also, maybe say something about the before/after images. For example, what you are going to do between them.

Author: I can add description. If you will write like this:
```
After:

![After](./images/after.png)
```
for markdown it will like you have one line - text and image will be on one line

Reviewer: If there's an empty line in Markdown in between, it should do a paragraph (HTML `<p> </p>`), no?

Author: only if two empty lines
otherwise, how do we split a long text that is limited to 150 characters into several lines and it saves the paragraph?

Reviewer: If you add a newline, it's still part of the same paragraph. The sentence just continues.

If you add an empty line in between (like <-- here), then it's rendered as new paragraph.

Not exactly sure with images, but the text and image should definitely not be on the same line.

Add this dependency to Rust with the shell in the same directory as `Cargo.toml`.

```bash
cargo add rand
```

Add file `addon.rs` and import it in `lib.rs`:

```rust
// file: lib.rs
mod addon;
```

> Add the following imports at the beginning of the file

Reviewer: ```suggestion
Add the following imports at the beginning of the file:
```

```rust
use godot::{
    engine::{
        Button, EditorInspectorPlugin, EditorPlugin, EditorProperty, IEditorInspectorPlugin,
        IEditorPlugin, IEditorProperty,
    },
    global,
    prelude::*,
};
use rand::Rng;

> ```

Reviewer: Please use flat imports. Also, `godot::engine` has been renamed to `godot::classes`.

Reviewer: This is not addressed, please make sure only conversations are resolved that are actually fixed. It's a lot of effort if I have to double-check 20 parallel conversations.

Since Rust is a statically typed language, we will proceed in reverse order unlike in Godot documentation, to avoid encountering errors unnecessarily.

## Add Property Editor

To begin with, let's define the editor for properties:

```rust
#[derive(GodotClass)]
#[class(tool, init, base=EditorProperty)]
struct RandomIntEditor {
    base: Base<EditorProperty>,
    button: Option<Gd<Button>>,
}
```

After that, we need to add an implementation for the trait `IEditorProperty`:

```rust
#[godot_api]
impl IEditorProperty for RandomIntEditor {
    fn enter_tree(&mut self) {

>         // Create button element
>         let mut button = Button::new_alloc();
>         // Add handler for this button, handle_press will be define in another impl
>         button.connect(
>             "pressed".into(),
>             self.base().callable("handle_press".to_godot()),
>         );
>         button.set_text("Randomize".into());
>         // Save pointer to the button into struct
>         self.button = Some(button.clone());
>         self.base_mut().add_child(button.to_variant().to());

Reviewer: Code style:
- Comments should end with full stops.
- Before comments, add empty line, so it's clear what belongs together.
```suggestion
        // Create button element.
        let mut button = Button::new_alloc();
        
        // Add handler for this button, handle_press will be define in another impl.
        button.connect(
            "pressed".into(),
            self.base().callable("handle_press".to_godot()),
        );
        button.set_text("Randomize".into());
        
        // Save pointer to the button into struct.
        self.button = Some(button.clone());
        self.base_mut().add_child(button.to_variant().to());
```

            let text = format!("Randomize: {}", num);
            button.set_text(text.into());
        } else {
            // Print error of something went wrong
            godot_error!("Button wasn't found in handle_press");
        }
    }
}
```

## Add Inspector plugin

Now we need to connect this editor to fields with an integer type.
To do this, we need to create an EditorInspectorPlugin.

```rust
#[derive(GodotClass)]
#[class(tool, init, base=EditorInspectorPlugin)]
struct RandomInspectorPlugin {
    base: Base<EditorInspectorPlugin>,
}
```

IEditorInspectorPlugin implementation:

```rust
#[godot_api]
impl IEditorInspectorPlugin for RandomInspectorPlugin {
      fn parse_property(
        &mut self,
        _object: Gd<Object>, // object that inspecting
        value_type: VariantType,
        name: GString,
        _hint_type: global::PropertyHint,
        _hit_name: GString,
        _flags: global::PropertyUsageFlags,
        _wide: bool,
    ) -> bool {
        if value_type == VariantType::INT {
            self.base_mut()
                .add_property_editor(name, RandomIntEditor::new_alloc().to_variant().to());
            return true;
        }

        false

>     }

Reviewer: Also here, I'd use the official names (see earlier comment).

And the `to_variant().to()` can be `upcast()`.

    fn can_handle(&self, _object: Gd<Object>) -> bool {
        true

>     }

Reviewer: Do you have an explanation (maybe as a comment) why this is always true?

Author: I'll add for node2d, and tell what is about

}
```

If `parse_property` returns `true`, the editor will be created and replace the current
representation; if not, it's necessary to return `false`.
This allows for specific control over where and how processing occurs.

> ## Add Editor Plugin

Reviewer: ```suggestion
## Adding an editor plugin
```

Only one thing left to do: define the editor plugin that will kick off all this magic!
This can be a generic `EditorPlugin` or a more specific `InspectorEditorPlugin`, depending
on what you want to achieve.

```rust
#[derive(GodotClass)]
#[class(tool, init, editor_plugin, base=EditorPlugin)]
struct RustEditorPlugin {
    base: Base<EditorPlugin>,
    random_inspector: Gd<RandomInspectorPlugin>,
}
```

```rust
#[godot_api]
impl IEditorPlugin for RustEditorPlugin {
    fn enter_tree(&mut self) {
        // create our inspector plugin and save it to next remove
        self.random_inspector = RandomInspectorPlugin::new_gd();
        self.to_gd()
            .add_inspector_plugin(self.random_inspector.to_variant().to());
    }

    fn exit_tree(&mut self) {
        self.to_gd()
            .remove_inspector_plugin(self.random_inspector.to_variant().to());

>     }

Reviewer: See [above](https://github.com/godot-rust/book/pull/48/files#r1632296129).

}
```

```admonish danger title="Troubleshooting"
Sometimes after compilation, you may encounter errors or panic. Most likely, all you need to do is simply **restart** the Godot Editor.

> ```

Reviewer: Could be admonish `note`. Warnings/dangers are not to explain what errors mean to the user, but to make them aware of potential pitfalls.

   at: (core/object/class_db.cpp:392)
ERROR: Cannot get class 'RandomInspectorPlugin'.
   at: (core/object/class_db.cpp:392)
```

# Full code

```rust
// file: addon.rs

use godot::{
    engine::{
        Button, EditorInspectorPlugin, EditorPlugin, EditorProperty, IEditorInspectorPlugin,
        IEditorPlugin, IEditorProperty,
    },
    global,
    prelude::*,
};
use rand::Rng;

#[derive(GodotClass)]
#[class(tool, init, editor_plugin, base=EditorPlugin)]
struct RustEditorPlugin {
    base: Base<EditorPlugin>,
    random_inspector: Gd<RandomInspectorPlugin>,
}

#[godot_api]
impl IEditorPlugin for RustEditorPlugin {
    fn enter_tree(&mut self) {
        self.random_inspector = RandomInspectorPlugin::new_gd();
        self.to_gd()
            .add_inspector_plugin(self.random_inspector.to_variant().to());
    }

    fn exit_tree(&mut self) {
        self.to_gd()
            .remove_inspector_plugin(self.random_inspector.to_variant().to());
    }
}

#[derive(GodotClass)]
#[class(tool, init, base=EditorInspectorPlugin)]
struct RandomInspectorPlugin {
    base: Base<EditorInspectorPlugin>,
}

#[godot_api]
impl IEditorInspectorPlugin for RandomInspectorPlugin {

>     fn parse_property(
>         &mut self,
>         _object: Gd<Object>,
>         value_type: VariantType,
>         name: GString,
>         _hint_type: global::PropertyHint,
>         _hit_name: GString,
>         _flags: global::PropertyUsageFlags,
>         _wide: bool,
>     ) -> bool {

Reviewer: Let's stick to the official parameter names (for `value_type` it's OK since `type` is a keyword):
```suggestion
    fn parse_property(
        &mut self,
        _object: Gd<Object>,
        value_type: VariantType,
        name: GString,
        _hint_type: global::PropertyHint,
        _hint_string: GString,
        _usage_flags: global::PropertyUsageFlags,
        _wide: bool,
    ) -> bool {
```

Author: ok, but what is "hint string" - i don't really undestand this suffix. looks redundant

Reviewer: It's a common term in Godot jargon. Admittedly low-level, but it appears in many places.

        if value_type == VariantType::INT {

>             self.base_mut()
>                 .add_property_editor(name, RandomIntEditor::new_alloc().to_variant().to());

Reviewer: ```suggestion
            self.base_mut()
                .add_property_editor(name, RandomIntEditor::new_alloc().upcast());
```

            return true;
        }

        false
    }

    fn can_handle(&self, _object: Gd<Object>) -> bool {
        true
    }
}

#[derive(GodotClass)]
#[class(tool, init, base=EditorProperty)]
struct RandomIntEditor {
    base: Base<EditorProperty>,
    button: Option<Gd<Button>>,
}

#[godot_api]
impl RandomIntEditor {

>     #[func]
>     fn handle_press(&mut self) {
>         let property_name = self.base().get_edited_property();
>         let num = rand::thread_rng().gen_range(0..100);
>         godot_print!("Randomize! {} for {}", num, property_name);
>         self.base_mut()
>             .emit_changed(property_name, num.to_variant().to());
>         if let Some(mut button) = self.button.clone() {
>             let text = format!("Randomize: {}", num);
>             button.set_text(text.into());
>         } else {
>             godot_error!("Button wasn't found in handle_press");
>         }
>     }

Reviewer: ```suggestion
    #[func]
    fn handle_press(&mut self) {
        let property_name = self.base().get_edited_property();
        let num = rand::thread_rng().gen_range(0..100);
        godot_print!("Randomize {} for {}", num, property_name);
        
        self.base_mut()
            .emit_changed(property_name, num.to_variant());
            
        if let Some(mut button) = self.button.clone() {
            let text = format!("Randomize: {num}");
            button.set_text(text.into());
        } else {
            godot_error!("Button wasn't found in handle_press");
        }
    }
```

}

#[godot_api]
impl IEditorProperty for RandomIntEditor {
    fn enter_tree(&mut self) {
        let mut button = Button::new_alloc();
        button.connect(
            "pressed".into(),

>             self.base().callable("handle_press".to_godot()),

Reviewer: ```suggestion
            self.base().callable("handle_press"),
```
Is this working? Because `base` method exists on `WithBaseField`...

Author: yes, this works

<!--
  ~ Copyright (c) godot-rust; Bromeon and contributors.
  ~ This Source Code Form is subject to the terms of the Mozilla Public
  ~ License, v. 2.0. If a copy of the MPL was not distributed with this
  ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->

# Inspector plugins

The inspector dock allows you to create custom widgets to edit properties through plugins.
This can be beneficial when working with custom datatypes and resources, although you can
use the feature to change the inspector widgets for built-in types. You can design custom
controls for specific properties, entire objects, and even separate controls associated

> with particular datatypes.

Reviewer 2: "For more info, see <https://docs.godotengine.org/en/stable/classes/class_editorinspectorplugin.html#class-editorinspectorplugin>."

Reviewer 2: If you want to put a description of the link, maybe it can be more descriptive, like "the `EditorInspectorPlugin` docs" or some such

This is implementation of

> [Godot docs](https://docs.godotengine.org/en/stable/tutorials/plugins/editor/inspector_plugins.html) on Rust.

Reviewer 2: ```suggestion
the [example](https://docs.godotengine.org/en/stable/tutorials/plugins/editor/inspector_plugins.html) in the Godot docs in Rust.
```

Before:

![Before](./images/before.png)

After:

![After](./images/after.png)

> Add requires to Rust (in Rust folder, where is Cargo.toml):

Reviewer 2: ```suggestion
Add this dependency to Rust with the shell in the same directory as `Cargo.toml`.
```

```bash
cargo add rand
```

Add file `addon.rs` and import it in `lib.rs`:

```rust
// file: lib.rs
mod addon;
```

Add the following imports at the beginning of the file

```rust
use godot::{
    classes,
    engine::{
        Button, EditorInspectorPlugin, EditorPlugin, EditorProperty, IEditorInspectorPlugin,
        IEditorPlugin, IEditorProperty,
    },
    global,
    prelude::*,
};
use rand::Rng;
```

> Since Rust is a statically typed language, we will proceed in reverse order unlike in Godot documentation, to avoid encountering errors unnecessarily.

Reviewer 2: Not sure this line is necessary since people going through this tutorial probably won't be going through the Godot one at the same time.  But you can leave it if you prefer.

Author: Just for show why it has upside down flow)

## Add Property Editor

To begin with, let's define the editor for properties:

```rust
#[derive(GodotClass)]
#[class(tool, init, base=EditorProperty)]

> struct RandomNumberEditor {

Reviewer 2: In the Godot example, the script is called `random_int_editor.gd`, so maybe it'd be better to call it `RandomIntEditor` to mirror their docs closely

        // - send prop name and random number to godot engine for update value
        // - update button text
        let property_name = self.base().get_edited_property();
        let num = rand::thread_rng().gen_range(0..100);
        godot_print!("Randomize! {} for {}", num, property_name);
        self.base_mut()
            .emit_changed(property_name, num.to_variant().to());
        if let Some(mut button) = self.button.clone() {
            let text = format!("Randomize: {}", num);
            button.set_text(text.into());
        } else {
            // Print error of something went wrong
            godot_error!("Button wasn't found in handle_press");
        }
    }
}
```

## Add Inspector plugin

Now we need to connect this editor to fields with an integer type.
To do this, we need to create an EditorInspectorPlugin.

```rust
#[derive(GodotClass)]
#[class(tool, init, base=EditorInspectorPlugin)]
struct RandomInspectorPlugin {
    base: Base<EditorInspectorPlugin>,
}
```

IEditorInspectorPlugin implementation:

```rust
#[godot_api]
impl IEditorInspectorPlugin for RandomInspectorPlugin {
    fn parse_property(
        &mut self,

>         _: Gd<classes::Object>, // inspect object

Reviewer 2: ```suggestion
        _: Gd<Object>, // inspect object
```
`Object` is in the prelude

        type_: VariantType, // value type
        name: GString, // property name
        _: global::PropertyHint, // hint type
        _: GString, // hit name
        _: global::PropertyUsageFlags, // property usage flag

>         _: bool, // wide

Reviewer 2: If you prefix the unused argument names with an underscore, then it doesn't trigger the unused variable lint and you won't need the comments, e.g. `_object: Gd<Object>`

Author: oh, I didn't know about, thanks!)

    ) -> bool {
        if type_ == VariantType::INT {
            let editor = RandomNumberEditor::new_alloc().to_variant().to();
            self.base_mut().add_property_editor(name, editor);
            return true;
        }

        false
    }

    fn can_handle(&self, _: Gd<classes::Object>) -> bool {
        true
    }
}
```

If `parse_property` returns `true`, the editor will be created and replace the current
representation; if not, it's necessary to return `false`.
This allows for specific control over where and how processing occurs.

## Add Editor Plugin

Only one thing left to do: define the editor plugin that will kick off all this magic!

> This can be a generic EditorPlugin or a more specific InspectorEditorPlugin, depending

Reviewer 2: ```suggestion
This can be a generic `EditorPlugin` or a more specific `InspectorEditorPlugin`, depending
```

on what you want to achieve.

```rust
#[derive(GodotClass)]
#[class(tool, init, editor_plugin, base=EditorPlugin)]
struct RustEditorPlugin {
    base: Base<EditorPlugin>,
    random_inspector: Gd<RandomInspectorPlugin>,
}
```

```rust
#[godot_api]
impl IEditorPlugin for RustEditorPlugin {
    fn enter_tree(&mut self) {
        // create our inspector plugin and save it to next remove
        self.random_inspector = RandomInspectorPlugin::new_gd();

>         self.to_gd()

Reviewer 2: ```suggestion
        self.base_mut()
```

Author: then I'll get:
```
error[E0502]: cannot borrow `self.random_inspector` as immutable because it is also borrowed as mutable
  --> src\addon.rs:23:35
   |
22 |         self.base_mut()
   |         ---------------
   |         |
   |         mutable borrow occurs here
   |         a temporary with access to the mutable borrow is created here ...
23 |             .add_inspector_plugin(self.random_inspector.to_variant().to());
   |                                   ^^^^^^^^^^^^^^^^^^^^^                   - ... and the mutable borrow might be used here, when that temporary is dropped and runs the destructor for type `BaseMut<'_, RustEditorPlugin>`
   |                                   |
   |                                   immutable borrow occurs here

```

Reviewer: You could create the argument in a separate variable, then it should work.

Author: yes, but what the difference? could you explain cons and pros?

Reviewer 3: `base()/base_mut()` preserve type safety, and can often avoid runtime errors that `to_gd()` wont be able to catch. The simplest example is this:
```rs
fn ready(&mut self) {
  // this will error at runtime because `on_notification` tries to get another mutable reference to self
  self.to_gd().add_child(Node::new());
  // whereas this wont because `base_mut` avoids that problem
  self.base_mut().add_child(Node::new());
}

fn on_notification(&mut self, what: NodeNotification) {

}
```

so whenever possible, you should prefer `base/base_mut` over `to_gd`

Author: thank you

        self.to_gd()
            .remove_inspector_plugin(self.random_inspector.to_variant().to());
    }
}
```

```admonish danger title="Troubleshooting"
Sometimes after compilation, you may encounter errors or panic. Most likely, all you need to do is simply **restart** the Godot Editor.
```

Example error:

```log
Initialize godot-rust (API v4.2.stable.official, runtime v4.2.2.stable.official)
ERROR: Cannot get class 'RandomInspectorPlugin'.
   at: (core/object/class_db.cpp:392)
ERROR: Cannot get class 'RandomInspectorPlugin'.
   at: (core/object/class_db.cpp:392)
```

# Full code

```rust
// file: addon.rs

use godot::{
    classes,
    engine::{
        Button, EditorInspectorPlugin, EditorPlugin, EditorProperty, IEditorInspectorPlugin,
        IEditorPlugin, IEditorProperty,
    },
    global,
    prelude::*,
};
use rand::Rng;

#[derive(GodotClass)]
#[class(tool, init, editor_plugin, base=EditorPlugin)]
struct RustEditorPlugin {
    base: Base<EditorPlugin>,
    random_inspector: Gd<RandomInspectorPlugin>,
}

#[godot_api]
impl IEditorPlugin for RustEditorPlugin {
    fn enter_tree(&mut self) {
        self.random_inspector = RandomInspectorPlugin::new_gd();

>         self.to_gd()

Reviewer 2: ```suggestion
        self.base_mut()
```

<!--
  ~ Copyright (c) godot-rust; Bromeon and contributors.
  ~ This Source Code Form is subject to the terms of the Mozilla Public
  ~ License, v. 2.0. If a copy of the MPL was not distributed with this
  ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->

# Inspector plugins

The inspector dock allows you to create custom widgets to edit properties through plugins.
This can be beneficial when working with custom datatypes and resources, although you can
use the feature to change the inspector widgets for built-in types. You can design custom
controls for specific properties, entire objects, and even separate controls associated
with particular datatypes.
For more info, see
[docs.godotengine.org](https://docs.godotengine.org/en/stable/classes/class_editorinspectorplugin.html#class-editorinspectorplugin).

The [example](https://docs.godotengine.org/en/stable/tutorials/plugins/editor/inspector_plugins.html)

> in the `Godot` docs in `Rust`. It will replace Integer input with button that randomize value.

Reviewer: This still misses the verb. And Godot/Rust are names, not code symbols, so they don't need to be in backticks.

Reviewer: > It will replace Integer input with button that randomize value.

Should be:
> It will replace integer input with a button that creates a random value.

Author: I apologize, just I am not native speaker

impl IEditorProperty for RandomIntEditor {
    fn enter_tree(&mut self) {
        // Create button element.
        let mut button = Button::new_alloc();
        
        // Add handler for this button, handle_press will be define in another impl.
        button.connect("pressed".into(), self.base().callable("handle_press"));
        button.set_text("Randomize".into());
        
        // Save pointer to the button into struct.
        self.button = Some(button.clone());
        self.base_mut().add_child(button.upcast());
    }

    fn exit_tree(&mut self) {
        // Remove element from inspector when this plugin unmount:
        if let Some(button) = self.button.take() {
            self.base_mut().remove_child(button.upcast());
        } else {
            // Log error if button disappeared before
            godot_error!("Button wasn't found in exit_tree");
        }
    }
}
```

Let's add a handler for the button:

```rust
#[godot_api]
impl RandomIntEditor {
    #[func]
    fn handle_press(&mut self) {
        // Update value by button click
        // - take prop name, randomize number
        // - send prop name and random number to godot engine for update value

>         // - update button text

Reviewer: Also here, multiple things:
- Sentence and list items start with uppercase
- Full stop . at end
- "Godot" capitalized
- "for update value" -> "to update value"
- "prop" -> "property"

Author: I am not sure about `Full stop . at end`, I've added dot at the end of every list items, right?

        let property_name = self.base().get_edited_property();
        let num = rand::thread_rng().gen_range(0..100);

>         godot_print!("Randomize! {} for {}", num, property_name);

Reviewer: ```suggestion
        godot_print!("Randomize: {num} for {property_name}");
```

        self.base_mut()

>             .emit_changed(property_name, num.to_variant().to());

Reviewer: Please fix this everywhere :slightly_smiling_face:

        if let Some(mut button) = self.button.clone() {

>             let text = format!("Randomize: {}", num);

Reviewer: ```suggestion
            let text = format!("Randomize: {num}");
```

            button.set_text(text.into());
        } else {
            // Print error of something went wrong
            godot_error!("Button wasn't found in handle_press");
        }
    }
}
```

## Add Inspector plugin

Now we need to connect this editor to fields with an integer type.

> To do this, we need to create an EditorInspectorPlugin.

Reviewer: ```suggestion
To do this, we need to create an `EditorInspectorPlugin`.
```

```rust
#[derive(GodotClass)]
#[class(tool, init, base=EditorInspectorPlugin)]
struct RandomInspectorPlugin {
    base: Base<EditorInspectorPlugin>,
}
```

> IEditorInspectorPlugin implementation:

Reviewer: Maybe write a short sentence, introducing the following code block.

```rust
#[godot_api]
impl IEditorInspectorPlugin for RandomInspectorPlugin {
      fn parse_property(
        &mut self,
        _object: Gd<Object>, // object that inspecting
        value_type: VariantType,
        name: GString,
        _hint_type: global::PropertyHint,
        _hit_string: GString,
        _flags: global::PropertyUsageFlags,
        _wide: bool,
    ) -> bool {
        if value_type == VariantType::INT {
            self.base_mut()
                .add_property_editor(name, RandomIntEditor::new_alloc().upcast());
            return true;
        }

        false
    }

    fn can_handle(&self, _object: Gd<Object>) -> bool {
        // 
        object.is_class("Node2D".into())

>     }

Reviewer: Did you plan to add a comment here? :wink:

Author: omg, how i missed this, even with broken example

}
```

If `parse_property` returns `true`, the editor will be created and replace the current
representation; if not, it's necessary to return `false`.

> This allows for specific control over where and how processing occurs.

Reviewer: > the editor will be created

do you mean "editor plugin" here?

> This allows for specific control over where and how processing occurs.

Processing what?

(sorry if I'm nitpicky, but it's important to be precise)

Author: I'll replace on `This allows you to control where and how processing is done by this plugin.`

## Adding an editor plugin

Only one thing left to do: define the editor plugin that will kick off all this magic!
This can be a generic `EditorPlugin` or a more specific `InspectorEditorPlugin`, depending
on what you want to achieve.

```rust
#[derive(GodotClass)]
#[class(tool, init, editor_plugin, base=EditorPlugin)]
struct RustEditorPlugin {
    base: Base<EditorPlugin>,
    random_inspector: Gd<RandomInspectorPlugin>,
}
```

```rust
#[godot_api]
impl IEditorPlugin for RustEditorPlugin {
    fn enter_tree(&mut self) {
        // create our inspector plugin and save it to next remove
        let plugin = RandomInspectorPlugin::new_gd();
        self.random_inspector = plugin.clone();
        self.base_mut().add_inspector_plugin(plugin.upcast());
    }

    fn exit_tree(&mut self) {
        // remove inspector plugin when editor plugin unmount
        let plugin = self.random_inspector.clone();
        self.base_mut().remove_inspector_plugin(plugin.upcast());

>     }

Reviewer: ```suggestion
    fn enter_tree(&mut self) {
        // Create our inspector plugin and save it.
        let plugin = RandomInspectorPlugin::new_gd();
        self.random_inspector = plugin.clone();
        self.base_mut().add_inspector_plugin(plugin.upcast());
    }

    fn exit_tree(&mut self) {
        // Remove inspector plugin when editor plugin leaves scene tree.
        let plugin = self.random_inspector.clone();
        self.base_mut().remove_inspector_plugin(plugin.upcast());
    }
```

}
```

```admonish note title="Troubleshooting"
Sometimes after compilation, you may encounter errors or panic. Most likely, all you need to do is simply **restart** the Godot Editor.
```

Example error:

```log
Initialize godot-rust (API v4.2.stable.official, runtime v4.2.2.stable.official)
ERROR: Cannot get class 'RandomInspectorPlugin'.
   at: (core/object/class_db.cpp:392)
ERROR: Cannot get class 'RandomInspectorPlugin'.
   at: (core/object/class_db.cpp:392)
```

> # Full code

Reviewer: I'm not sure about this -- it's nice to heave readily available code, but also quite hard to keep the two things in sync. In fact they're already out-of-sync now.

Probably over time, we could look into tooling that references snippets of files.

For now, I'd say there are two options:
- Remove the "full code" part
- Keep it, but make sure it 100% matches the code in the above snippets

Reviewer: Also, some of my comments were only done in either the "full code" section, or the part before -- could you see if they're all applied to the part before as well?

Author: yes, I did it in my file and this is mirror of it, but without comments. Yes, this is hard to sync. I'll remove it.

```rust
// file: addon.rs

use godot::{
    classes::{
        Button, EditorInspectorPlugin, EditorPlugin, EditorProperty, IEditorInspectorPlugin,
        IEditorPlugin, IEditorProperty,
    },
    global,
    prelude::*,

> };

Reviewer: Please also flat imports here :slightly_smiling_face:

```rust
#[godot_api]
impl IEditorInspectorPlugin for RandomInspectorPlugin {
      fn parse_property(
        &mut self,

>         _object: Gd<Object>, // object that inspecting

Reviewer 2: ```suggestion
        _object: Gd<Object>, // object that is being inspected
```
